File: app/src/main/java/com/mindfulai/customclass/FCMService.java
package com.mindfulai.customclass;

import android.annotation.SuppressLint;
import android.app.Notification;
import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.content.Intent;
import android.content.SharedPreferences;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Color;
import android.media.RingtoneManager;
import android.os.Build;
import android.util.Log;

import androidx.core.app.NotificationCompat;

import com.google.firebase.messaging.FirebaseMessagingService;
import com.google.firebase.messaging.RemoteMessage;
import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;
import com.mindfulai.Activites.MainActivity;
import com.mindfulai.Models.NotificationModel;
import com.mindfulai.Utils.GlobalEnum;
import com.mindfulai.ministore.R;

import java.io.InputStream;
import java.lang.reflect.Type;
import java.net.HttpURLConnection;
import java.net.URL;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;

public class FCMService extends FirebaseMessagingService {
    public static String tag = null;
    String type;
    String Id = "fcm_default_channel";
    Intent intent;
    private ArrayList<NotificationModel> notificationModelArrayList;

    @Override
    public void onCreate() {
        super.onCreate();
    }

    @SuppressLint("ResourceAsColor")
    @Override
    public void onMessageReceived(RemoteMessage remoteMessage) {
        loadData(remoteMessage);

        tag = remoteMessage.getData().get("tag");
        type = remoteMessage.getData().get("type");
        intent = new Intent(this, MainActivity.class);
        intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP);
        String imageUri = remoteMessage.getData().get("image");
        Bitmap bitmap = getBitmapfromUri(GlobalEnum.AMAZON_URL+imageUri);
        PendingIntent pendingIntent = PendingIntent.getActivity(this, 0, intent, PendingIntent.FLAG_UPDATE_CURRENT);
        NotificationCompat.Builder builder = new NotificationCompat.Builder(this, Id);
        if (bitmap != null) {
            builder.setStyle(new NotificationCompat.BigPictureStyle().bigPicture(bitmap).setBigContentTitle(remoteMessage.getData().get("title")).setSummaryText(remoteMessage.getData().get("body")));
        } else
            builder.setStyle(new NotificationCompat.BigTextStyle().bigText(remoteMessage.getData().get("body")).setBigContentTitle(remoteMessage.getData().get("title")));

        builder.setFullScreenIntent(pendingIntent, true);
        builder.setPriority(NotificationCompat.PRIORITY_MAX);
        builder.setAutoCancel(true);
        builder.setDefaults(Notification.DEFAULT_VIBRATE);
        builder.setVibrate(new long[]{10000, 10000, 10000, 10000, 10000});
        builder.setLights(Color.RED, 10000, 10000);
        builder.setSound(RingtoneManager.getDefaultUri(RingtoneManager.TYPE_NOTIFICATION));
        builder.setContentIntent(pendingIntent);
        NotificationManager manager = (NotificationManager) getSystemService(NOTIFICATION_SERVICE);
        builder.setSmallIcon(R.mipmap.ic_launcher);
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            NotificationChannel channel = new NotificationChannel(Id, "Default channel", NotificationManager.IMPORTANCE_HIGH);
            manager.createNotificationChannel(channel);
        }
        manager.notify(0, builder.build());
    }

    private Bitmap getBitmapfromUri(String imageUri) {
        URL url;
        HttpURLConnection connection;
        InputStream input;
        Bitmap bitmap = null;
        try {
            url = new URL(imageUri);
            connection = (HttpURLConnection) url.openConnection();
            connection.setDoInput(true);
            connection.connect();
            input = connection.getInputStream();
            bitmap = BitmapFactory.decodeStream(input);
        } catch (Exception e) {
            e.printStackTrace();
            Log.e("TAG", "getBitmapfromUri: " + e);
        }
        return bitmap;
    }
    private void saveData() {
        SharedPreferences sharedPreferences = getSharedPreferences("shared preferences", MODE_PRIVATE);
        SharedPreferences.Editor editor = sharedPreferences.edit();
        Gson gson = new Gson();
        String json = gson.toJson(notificationModelArrayList);
        editor.putString("notification list", json);
        editor.apply();
    }

    private void loadData(RemoteMessage remoteMessage) {
    SharedPreferences sharedPreferences = getSharedPreferences("shared preferences", MODE_PRIVATE);
    Gson gson = new Gson();
    String json = sharedPreferences.getString("notification list", "");
    Type type = new TypeToken<ArrayList<NotificationModel>>() {}.getType();
    notificationModelArrayList = gson.fromJson(json, type);
    if (notificationModelArrayList == null) {
        notificationModelArrayList = new ArrayList<>();
     }
    Map<String ,String> hashMap = remoteMessage.getData();

    if(hashMap.get("image")!=null)
       notificationModelArrayList.add(new NotificationModel(hashMap.get("tag"),hashMap.get("type"),hashMap.get("image"),hashMap.get("title"),hashMap.get("title")));
   else
       notificationModelArrayList.add(new NotificationModel(hashMap.get("tag"),hashMap.get("type"),"",hashMap.get("title"),hashMap.get("title")));

        saveData();
    }
}

File: app/src/main/java/com/mindfulai/Utils/FCMService.kt
package com.mindfulai.Utils;


import android.app.NotificationChannel
import android.app.NotificationManager
import android.content.Context
import android.os.Build
import android.util.Log
import androidx.annotation.RequiresApi
import androidx.core.app.NotificationCompat
import androidx.core.app.NotificationManagerCompat
import com.bumptech.glide.Glide
import com.google.firebase.messaging.FirebaseMessagingService
import com.google.firebase.messaging.RemoteMessage
import com.mindfulai.Models.Notification
import com.mindfulai.dao.AppDatabase
import com.mindfulai.ministore.R
import java.util.*


class FCMService : FirebaseMessagingService() {

    override fun onMessageReceived(message: RemoteMessage) {
        try {
            Log.i("FcmNotification", message?.notification.toString())
            Log.i("FcmData", message?.data.toString())
            createNotificationChannel(message)
            var builder = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
                getBuilder(message)
            } else {
                null
            }
            with(NotificationManagerCompat.from(this)) {
                builder?.build()?.let { notify(2, it) }
                Log.i("fcmNotification", "sent")
            }
            saveNotification(message)
        } catch (e: Exception) {
            e.printStackTrace()
        }

    }


    private fun saveNotification(message: RemoteMessage) {
        var notification = Notification(
                Math.random().toLong(),
                message.data["title"].toString(),
                message.data["body"].toString(),
                message.data["image"] ?: "",
                Date().toString())
        AppDatabase.getDatabase(applicationContext)?.notificationDao()?.saveNotification(notification)
        Log.i("notifications##", AppDatabase.getDatabase(applicationContext)?.notificationDao()?.getAllNotifications().toString() + "####")
    }


    @RequiresApi(Build.VERSION_CODES.LOLLIPOP)
    private fun getBuilder(message: RemoteMessage): NotificationCompat.Builder? {
        var builder: NotificationCompat.Builder? = null
        builder = message?.data?.get("channel_id")?.let {
            NotificationCompat.Builder(this, it)
                    .setSmallIcon(R.drawable.granny_logo)
                    .setContentTitle(message?.data?.get("title"))
                    .setContentText(message?.data?.get("body"))
                    .setStyle(NotificationCompat.BigTextStyle()
                            .bigText(message?.data?.get("body")))
                    .setPriority(NotificationCompat.PRIORITY_DEFAULT)
                    .setLargeIcon(Glide.with(this)
                            .asBitmap()
                            .load(R.drawable.ic_notification)
                            .submit()
                            .get())
        }
        if (message.data?.get("image") != null) {
            builder?.setLargeIcon(Glide.with(this)
                    .asBitmap()
                    .load(GlobalEnum.AMAZON_URL+message?.data?.get("image"))
                    .submit()
                    .get())
                    ?.setStyle(NotificationCompat.BigPictureStyle()
                            .bigPicture(Glide.with(this)
                                    .asBitmap()
                                    .load(GlobalEnum.AMAZON_URL+message?.data?.get("image"))
                                    .submit()
                                    .get())
                            .bigLargeIcon(null))
        }
        return builder
    }

    private fun createNotificationChannel(message: RemoteMessage) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            val importance = NotificationManager.IMPORTANCE_DEFAULT
            val channel = NotificationChannel(
                    message?.data?.get("channel_id"),
                    CommonUtils.capitalizeWord(message?.data?.get("channel_id")),
                    importance).apply {
                description = "Push notifications"
            }
            val notificationManager: NotificationManager =
                    getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
            notificationManager.createNotificationChannel(channel)
        }
    }


    override fun onNewToken(token: String) {
        super.onNewToken(token)
        Log.e("TAG", token)
    }
}
